### DISCUSSION TITLE
How to execute python code on mystjs markdown file?

### DISCUSSION THREAD

**Discussion by fcarvalhopacheco** [executablebooks/meta#1037] [Q&A]
@rowanc1 
Can I execute a python code inside a mystjs markdown file?

I have `test.md` with something like the following but it doest work:

````
some text... 
```{code-cell} python
import pandas as pd
from IPython.display import display, HTML


def read_file(file_path):
     # Read the file
    with open(file_path, 'r') as file:
        lines = file.readlines()
    
    # Remove header lines starting with '%'
    data_lines = [line.strip() for line in lines if not line.startswith('%')]
    
    # Split the lines into columns
    columns = ['Date', 'Time', 'Batch', '2xK-15', '2xC(1)', '2xC(2)', '2xC', 'Salinity', 'SKS', 'Air Temp', 'NaN']
    data = [line.split() for line in data_lines]
    
    # Create a DataFrame
    df = pd.DataFrame(data, columns=columns)
    
    return df

# Provide the file path
file_path = '../data/orig/IAPSO339.dat'

# Call the function to read the file
df = read_file(file_path)

display(df)

{eval}`display(df)`
```

````
I have tried to add the following on my `myst.yml`

```` 
...
project:
  title: HOT-{{cookiecutter.cruise_id}}
  # description:
  thebe:
    lite: true
...
````

I have also tried to add the following on the top of `test.md`:

```
---
jupytext:
  formats: md:myst
  text_representation:
    extension: .md
    format_name: myst
kernelspec:
  display_name: Python 3
  language: python
  name: python3
title: Autosal Standardization
description: Autosal Standardization
subject: Chapter 1
numbering:
  enumerator: 1.%s
export: 
  - format: pdf
    template: arxiv_nips
  - output: _build/export/autosal.pdf
---
```

I can only execute the code on `.ipynb`  but I cant combine multiple files...

**Comment by rowanc1** [comment]
Haven't tried this combination before yet, and can confirm that it is not working locally, but also it is _very_ close to working! There was a bug that @stevejpurves fixed that does require new templates (you can call `myst clean --templates` to get the fresh ones). With your setup (including the frontmatter for the markdown notebook) I can get as far as about here:

![image](https://github.com/executablebooks/meta/assets/913249/8dee45d7-435c-42b1-943d-1122f08aea5b)

Thebe loads, and executes the cells, but nothing is shown. @stevejpurves my guess is that we are not processing the `code-cell` into a block and an output when processing markdown - and then `thebe` is not able to attach itself to any output area.

If I put this node manually into the `mdast`:
```
          {
            "type": "output",
            "data": []
          }
```

Then it works with JupyterLite and matplotlib, etc. which is pretty cool actually. :)

![image](https://github.com/executablebooks/meta/assets/913249/5914c344-6280-4a7c-a84a-78aed56960ff)

This allows anyone in the browser to execute the markdown notebook, I think that might be what you mean @fcarvalhopacheco? But you also might be wanting to pre-execute using the CLI? The pre-execute/cache is something that is not yet complete.

**Reply by fcarvalhopacheco** [reply]
Thanks @rowanc1.

I am not sure if I understood your answer. 
It seems that you are using a ```.ipynb``` instead of ```.md``` file right? what/where is the ```mdast```?

I can't execute any code using ```.md``` files.

**Reply by rowanc1** [reply]
Sorry. :) I **was** using a markdown file!

The markdown gets parsed into a json file in `_build/site` (which is a markdown abstract syntax tree, or mdast) and these are mostly my notes to @stevejpurves for what we need to do to get this working. There is a missing node in the tree, which means `thebe` can't connect to anything, and no outputs are shown.

Other than that bug, all the infrastructure is there to create markdown files and have them be run (using JupyterLite) in the web-browser.

**Reply by rowanc1** [reply]
Tracking the bug here:
https://github.com/executablebooks/mystjs/issues/439
